snow/engine/snowman/bootstrap/storage.go

```diff
@@ -0,0 +1,252 @@
+// (lastAcceptedHeight, highestTrackedHeight].
+//
+// For example, if the tree currently contains heights [1, 4, 6, 7] and the
+// lastAcceptedHeight is 2, this function will return the IDs corresponding to
+// blocks [3, 5].
+func getMissingBlockIDs(
+	ctx context.Context,
+	db database.KeyValueReader,
+	parser block.Parser,
+	tree *interval.Tree,
+	lastAcceptedHeight uint64,
+) (set.Set[ids.ID], error) {
+	var (
+		missingBlocks     set.Set[ids.ID]
+		intervals         = tree.Flatten()
+		lastHeightToFetch = lastAcceptedHeight + 1
+	)
+	for _, i := range intervals {
+		if i.LowerBound <= lastHeightToFetch {
+			continue
+		}
+
+		blkBytes, err := interval.GetBlock(db, i.LowerBound)
+		if err != nil {
+			return nil, err
+		}
+
+		blk, err := parser.ParseBlock(ctx, blkBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		parentID := blk.Parent()
+		missingBlocks.Add(parentID)
+	}
+	return missingBlocks, nil
+}
+
+// process a series of consecutive blocks starting at [blk].
+//
+//   - blk is a block that is assumed to have been marked as acceptable by the
+//     bootstrapping engine.
+//   - ancestors is a set of blocks that can be used to lookup blocks.
+//
+// If [blk]'s height is <= the last accepted height, then it will be removed
+// from the missingIDs set.
+//
+// Returns a newly discovered blockID that should be fetched.
+func process(
+	db database.KeyValueWriterDeleter,
+	tree *interval.Tree,
+	missingBlockIDs set.Set[ids.ID],
+	lastAcceptedHeight uint64,
+	blk snowman.Block,
+	ancestors map[ids.ID]snowman.Block,
+) (ids.ID, bool, error) {
+	for {
+		// It's possible that missingBlockIDs contain values contained inside of
+		// ancestors. So, it's important to remove IDs from the set for each
+		// iteration, not just the first block's ID.
+		blkID := blk.ID()
+		missingBlockIDs.Remove(blkID)
+
+		height := blk.Height()
+		blkBytes := blk.Bytes()
+		wantsParent, err := interval.Add(
+			db,
+			tree,
+			lastAcceptedHeight,
+			height,
+			blkBytes,
+		)
+		if err != nil || !wantsParent {
+			return ids.Empty, false, err
+		}
+
+		// If the parent was provided in the ancestors set, we can immediately
+		// process it.
+		parentID := blk.Parent()
+		parent, ok := ancestors[parentID]
+		if !ok {
+			return parentID, true, nil
+		}
+
+		blk = parent
+	}
+}
+
+// execute all the blocks tracked by the tree. If a block is in the tree but is
+// already accepted based on the lastAcceptedHeight, it will be removed from the
+// tree but not executed.
+//
+// execute assumes that getMissingBlockIDs would return an empty set.
+//
+// TODO: Replace usage of haltable with context cancellation.
+func execute(
+	ctx context.Context,
+	haltable common.Haltable,
+	log logging.Func,
+	db database.Database,
+	parser block.Parser,
+	tree *interval.Tree,
+	lastAcceptedHeight uint64,
+) error {
+	var (
+		batch                    = db.NewBatch()
+		processedSinceBatchWrite uint
+		writeBatch               = func() error {
```

Comment:
nit: not sure if it makes a difference but should we pull this function out to avoid creating writeBatch object at each execute invocation?